qiskit/circuit/library/arithmetic/weighted_adder.py

```diff
@@ -0,0 +1,324 @@
+
+    Given :math:`n` qubits :math:`q_0, \ldots, q_{n-1}` and non-negative integer weights
+    :math:`\lambda_0, \ldots, \lambda_{n-1}`, this circuit performs the operation
+
+    .. math::
+
+        |q_0 \ldots q_{n-1}\rangle |0\rangle_s
+        \mapsto |q_0 \ldots q_{n-1}\rangle |\sum_{j=0}^{n-1} \lambda_j q_j\rangle_s
+
+    where :math:`s` is the number of sum qubits required.
+    This can be computed as
+
+    .. math::
+
+        s = 1 + \left\lfloor \log_2\left( \sum_{j=0}^{n-1} \lambda_j \right) \right\rfloor
+
+    or :math:`s = 1` if the sum of the weights is 0 (then the expression in the logarithm is
+    invalid).
+
+    For qubits in a circuit diagram, the first weight applies to the upper-most qubit.
+    For an example where the state of 4 qubits is added into a sum register, the circuit can
+    be schematically drawn as
+
+    .. parsed-literal::
+
+                   ┌────────┐
+          state_0: ┤0       ├ \ state_0 * weights[0]
+                   │        │ |
+          state_1: ┤1       ├ | + state_1 * weights[1]
+                   │        │ |
+          state_2: ┤2       ├ | + state_2 * weights[2]
+                   │        │ |
+          state_3: ┤3       ├ / + state_3 * weights[3]
+                   │        │
+            sum_0: ┤4       ├ \
+                   │  Adder │ |
+            sum_1: ┤5       ├ | = sum_0 * 2^0 + sum_1 * 2^1 + sum_2 * 2^2
+                   │        │ |
+            sum_2: ┤6       ├ /
+                   │        │
+          carry_0: ┤7       ├
+                   │        │
+          carry_1: ┤8       ├
+                   │        │
+        control_0: ┤9       ├
+                   └────────┘
+    """
+
+    def __init__(self,
+                 num_state_qubits: Optional[int] = None,
+                 weights: Optional[List[int]] = None,
+                 name: str = 'adder') -> None:
+        """Computes the weighted sum controlled by state qubits.
+
+        Args:
+            num_state_qubits: The number of state qubits.
+            weights: The weights per state qubits.
+            name: The name of the circuit.
+        """
+        super().__init__(name=name)
+
+        self._weights = None
+        self._num_state_qubits = None
+
+        self.weights = weights
+        self.num_state_qubits = num_state_qubits
+
+    @property
+    def num_sum_qubits(self) -> int:
+        """The number of sum qubits in the circuit.
+
+        Returns:
+            The number of qubits needed to represent the weighted sum of the qubits.
+        """
+        if sum(self.weights) > 0:
+            return int(np.floor(np.log2(sum(self.weights))) + 1)
+        return 1
+
+    @property
+    def weights(self) -> List[int]:
+        """The weights for the qubit states.
+
+        Returns:
+            The weight for the qubit states.
+        """
+        if self._weights:
+            return self._weights
+        if self.num_state_qubits:
+            return [1] * self.num_state_qubits
```

Comment:
so if the user does not specify the `weights`, they are all assumed to be 1? This should be documented in the class as its behavior.